Updated Readme to add reference to ITA documentation. (#2)
arvind5/trustauthority-samples: deployment/README.md

There are four components required to build the demo app:

- Dockerfile to build the Intel® Key Broker Service (Intel KBS) relying party container.
- **KBS.env** environment file to configure Intel KBS. You must modify this file to suit your environment. 
- Dockerfile to build the Intel TDX workload container. The Intel TDX workload build process supports Intel SGX DCAP, Azure confidential VMs with Intel TDX, and Google Cloud Platform (GCP) Confidential VMs.
- **workload.env** to supply the workload with Intel KBS and Intel Trust Authority connection info. 

## Prerequisites
- [Docker engine >= 20.10](https://docs.docker.com/engine/install/)

## Intel KBS

This section describes how to build and deploy the Intel KBS relying party container. For this demo, Intel KBS uses [PyKMIP](https://github.com/OpenKMIP/PyKMIP) for the key management system (KMS). PyKMIP is a relatively simple, standalone KMS that is useful for development and testing, but it is not a secure, production-quality KMS. Don't use PyKMIP for storing actual secrets.

### Build

> 1. Download the [Dockerfile](./kbs/Dockerfile) for building the  container.

Reviewer: ```suggestion
1. Download the [Dockerfile](./kbs/Dockerfile) for building the container.
```

### Deploy

1. Download the [workload.env](./sample-workload/workload.env) file and update the missing configurations.
   ```bash
   wget https://raw.githubusercontent.com/arvind5/trustauthority-samples/task/dockerfile-deployment/deployment/sample-workload/workload.env
   ```
   The **workload.env** file contains the settings needed for the sample workload to communicate with Intel KBS and Intel Trust Authority. This file contains secrets. Embedding secrets in an unencrypted file is _not_ a recommended best practice for secure systems. It's done here to make the demo easy to configure.

   The `KBS_ADMIN` and `KBS_PASSWORD` must match the `ADMIN_USERNAME` and `ADMIN_PASSWORD` used in KBS.env.

   The `KBS_URL` is of the form `https://<IP_address>:9443/kbs/v1`, where IP_address is the IP address of the host machine where the relying-party (KBS) container is running. Port 9443 is the default listening port for Intel KBS. If you change the port number for the Intel KBS container, you'll need to make a corresponding change to `KBS_URL`.

   `TRUSTAUTHORITY_API_KEY` is the same as used in **kbs.env**.

   ```
   KBS_ADMIN=<KBS-admin-username>
   KBS_PASSWORD=<KBS-admin-password>

>    KBS_URL=<KBS-Url>

Reviewer: ```suggestion
KBS_URL=https://<KBS-IP>:9443/kbs/v1
SKIP_TLS_VERIFICATION=<true | false>
HTTPS_PROXY=<proxy URL if required>
```

Reviewer: I'm not sure what SKIP_TLS_VERIFICATION does, exactly.

Author: SKIP_TLS_VERIFICATION if set to true will skip the TLS server certificate verification. we are setting it to true since we are using self-signed certificates for TLS.

2. Run the following command to start Intel KBS as a Docker container.
   ```bash
   sudo docker run --name relying-party -d --env-file kbs.env --restart=always -p 9443:9443 relying-party:latest
   ```

   9443 is the port Intel KBS will listen on. The port can be changed, if necessary.

## Intel TDX Demo workload

This section describes how to build and deploy the demonstration workload. The workload is an application that runs a machine learning (ML) model in an Intel TDX trust domain. The ML model is assumed to contain confidential data and parameters, therefore it is encrypted when first deployed. The demo workload (attester) must obtain an Intel Trust Authority attestation token and include the token with a request to Intel KBS to retrieve the key needed to decrypt the ML model. This workflow implements a secure key release use case in passport attestation mode.

### Build
1. Download the [Dockerfile](./sample-workload/Dockerfile) for building the Intel TDX demo workload.
   ```bash

>    wget https://raw.githubusercontent.com/arvind5/trustauthority-samples/task/dockerfile-deployment/deployment/sample-workload/Dockerfile

Reviewer: Did you mean to use a link to your fork, or is this temporary?

Author: it should be replaced with public URL before push, let me make this change now since everything else is complete.

Author: updated.

### Deploy
1. Download the [workload.env](./sample-workload/workload.env) file and update the missing configurations.
   ```bash
   wget https://raw.githubusercontent.com/arvind5/trustauthority-samples/task/dockerfile-deployment/deployment/sample-workload/workload.env
   ```
   The **workload.env** file contains the settings needed for the sample workload to communicate with Intel KBS and Intel Trust Authority. This file contains secrets. Embedding secrets in an unencrypted file is _not_ a recommended best practice for secure systems. It's done here to make the demo easy to configure.

   The `KBS_ADMIN` and `KBS_PASSWORD` must match the `ADMIN_USERNAME` and `ADMIN_PASSWORD` used in KBS.env.

   The `KBS_URL` is of the form `https://<IP_address>:9443/kbs/v1`, where IP_address is the IP address of the host machine where the relying-party (KBS) container is running. If the KBS is running on same TDVM as the workload, then provide `172.17.0.1` as IP_address. Port 9443 is the default listening port for Intel KBS. If you change the port number for the Intel KBS container, you'll need to make a corresponding change to `KBS_URL`.

> 

Reviewer: ```suggestion
SKIP_TLS_VERIFICATION if set to **true** will skip the TLS server certificate verification. For this demo, set it to **true** since we are using self-signed certificates for TLS.
```

   `TRUSTAUTHORITY_API_KEY` is the same as used in **kbs.env**.

   ```
   KBS_ADMIN=<KBS-admin-username>
   KBS_PASSWORD=<KBS-admin-password>
   KBS_URL=https://<IP_address>:9443/kbs/v1
   SKIP_TLS_VERIFICATION=<true | false>
   HTTPS_PROXY=<proxy URL if required>
   TRUSTAUTHORITY_API_URL=https://api.trustauthority.intel.com
   TRUSTAUTHORITY_API_KEY=<API Key>
   ```
2. Run the following command to start the demo workload as a Docker container.
   ```bash
   sudo docker run --name ita-demo -d --env-file workload.env --device=/dev/tdx_guest -p 12780:12780 --user 0 trustauthority-demo:latest
   ```

>    On successful run, the container will generate `execute_workload_flow.env` file under /tmp/ folder. This env file will be used later for executing workload flow. Please update the env variables in execute_workload_flow.env file if you used custom settings.

Reviewer: ```suggestion
   On successful run, the container will generate a `execute_workload_flow.env` file in the `/tmp/` folder. This env file will be used later for executing the workload flow script. Update the env variables in the `execute_workload_flow.env` file if you used custom settings.
```

> [!NOTE]
> If running on Azure, replace `--device=/dev/tdx_guest` with `--device=/dev/tpmrm0`  
> The command above will start the container as root user. This is needed because the Intel TDX demo workload container requires elevated privileges to access `/dev/tdx_guest` or `/dev/tpmrm0` device for quote collection.  
> If you don't want to run the container as root user, make sure the `/dev/tdx_guest` or `/dev/tpmrm0` device is accessible to the container. One of the ways to do this is to pass `gid` owning the `tdx_guest` or `tpmrm0` device to container.  
> ```bash
> sudo docker run --name ita-demo -d --env-file workload.env --device=/dev/tdx_guest -p 12780:12780 --group-add $(getent group <user-group> | cut -d: -f3) trustauthority-demo:latest
> ```
> Where `<user-group>` is the group owning the `tdx_guest` or `tpmrm0` device
> ```bash
> crw-rw---- 1 root <user-group> 10, 123 /dev/tdx_guest
> ```

### Execute Workload Flow
1. Download the [execute_workload_flow.sh](./sample-workload/execute_workload_flow.sh) script for executing secure key release workflow.
   ```bash

>    wget https://raw.githubusercontent.com/arvind5/trustauthority-samples/task/dockerfile-deployment/deployment/sample-workload/execute_workload_flow.sh

Reviewer: Another link to your fork. Intentional?

Author: updated.
